lib/cdo/redshift.rb

```diff
@@ -16,6 +37,16 @@ def initialize
   end
 
   def exec(sql_query)
-    @conn.exec(sql_query)
+    result = @conn.exec(sql_query)
+    # result_status is an undocumented PG::Result method that returns the integer query status.
+    # res_status is an undocumented PG::Result method that returns the string code for a given integer query status.
+    # API - https://www.rubydoc.info/gems/pg/PG/Result
+    # Source - https://github.com/ged/ruby-pg/blob/master/lib/pg/result.rb
+    result_status = result.res_status(result.result_status)
+    unless (QUERY_SUCCESS_CODES + QUERY_WARNING_CODES).include? result_status
+      @conn.exec('ROLLBACK;')
```

Comment:
This added line seems suspicious to me. Does this add an assumption/requirement on the `exec` method that the `sql_query` either creates or is contained within a transaction? Did we cover this exception code in the manual-testing process, to confirm that this rolls back correctly as expected on error?

Reply:
PostgreSQL enters an ABORT STATE when a statement fails execution and all subsequent statements executed in the same client session (even if they're not in the transaction block where the exception was raised) fail until the ABORT STATE is cleared with a COMMIT or a ROLLBACK.  We share the same Redshift client connection through the Redshift singleton class, so I've defaulted to explicitly rolling back whatever transaction we're in to clear the abort state.

Would an optional argument to this `exec` method specifying that it should NOT rollback on failure be better?  Or should we require that anyone invoking this method be a good citizen and use PostgreSQL exception handlers to catch the exception and clear the abort state?

Reply:
I believe I discovered this problem during manual testing when a single table rename failed and all subsequent queries failed due to the abort state.

Reply:
oh ok, thanks for the extra context, I didn't know about this quirk in PostgreSQL. I think this is fine like this, I'll keep this quirk in mind and we can revisit if it ever causes any future issues.